### PR TITLE
op-challenger: Update Claim types

### PR DESCRIPTION
**Pull request by trianglesphere**
**Description**

This updates the claim type to focus on the claim data & to split out the claim data from the claim. This makes it easier to reference the parent claim data without pointers. Avoid pointers is helpful because it limits the chance for memory confusion and enforces cleaner boundaries by simply passing data around.

### CODE REVIEW

File: op-challenger/fault/types.go [hunk 1]
```diff
@@ -19,19 +19,31 @@ type TraceProvider interface {
...
+
+func (c *Claim) DefendsParent() bool {
+	return c.IndexAtDepth() != c.Parent.IndexAtDepth()
 }
 
 // Responder takes a response action & executes.
 // For full op-challenger this means executing the transaction on chain.
 type Responder interface {
-	Respond(ctx context.Context, response Response) error
+	Respond(ctx context.Context, response Claim) error
```
Comment (refcell) [inline op-challenger/fault/types.go:48]: We lose the explicit `Response.Attack` information by doing this, so the solver `NextMove` just spits out a `Claim`. Is that all we should need for the game solver (in particular `Agent`) to respond?

File: op-challenger/fault/types.go [hunk 2]
```diff
@@ -19,19 +19,31 @@ type TraceProvider interface {
...
+// If the position of the game is Depth 0, IndexAtDepth 0 it is the root claim
+// and the Parent field is empty & meaningless.
+type Claim struct {
+	ClaimData
+	Parent ClaimData
+}
+
+func (c *Claim) IsRoot() bool {
+	return c.Position.IsRootPosition()
+}
```
Comment (refcell) [inline op-challenger/fault/types.go:39]: This is very nice

File: op-challenger/fault/types.go (hunk 1 above)
Comment (trianglesphere) [reply op-challenger/fault/types.go:48]: Yea - the claim type now has `IsDefend` available.

File: op-challenger/fault/cmd/examples/solver.go [hunk 3]
```diff
@@ -8,76 +8,72 @@ import (
...
+		fmt.Printf("%s\ttrace %v letter %c\n", name, idx, letter)
+	} else {
+		fmt.Printf("%s\ttrace %v letter %c is attack %v\n", name, idx, letter, !claim.DefendsParent())
+	}
+
+}
+
 // SolverExampleOne uses the [fault.Solver] with a [fault.AlphabetProvider]
 // to print out fault game traces for the "abcdexyz" counter-state.
 func SolverExampleOne() {
```
Comment (refcell) [inline op-challenger/fault/cmd/examples/solver.go:25]: Getting this output when I run this:
```
➜  cmd git:(jg/updates_types) go run main.go
# github.com/ethereum-optimism/optimism/op-challenger/fault/cmd/examples
examples/solver.go:23:3: unknown field Value in struct literal of type fault.Claim
examples/solver.go:24:3: unknown field Position in struct literal of type fault.Claim
examples/solver.go:46:3: unknown field Value in struct literal of type fault.Claim
examples/solver.go:47:3: unknown field Position in struct literal of type fault.Claim
examples/solver.go:49:50: too many arguments in call to solver.NextMove
	have (fault.Claim, fault.Claim)
	want (fault.Claim)
examples/solver.go:64:3: unknown field Value in struct literal of type fault.Claim
examples/solver.go:65:3: unknown field Position in struct literal of type fault.Claim
examples/solver.go:67:49: too many arguments in call to solver.NextMove
	have (fault.Claim, fault.Claim)
	want (fault.Claim)
examples/solver.go:77:5: non-boolean condition in if statement
```

Comment (trianglesphere) [reply op-challenger/fault/cmd/examples/solver.go:25]: I just force pushed - working locally

File: op-challenger/fault/types.go (hunk 1 above)
Comment (refcell) [reply op-challenger/fault/types.go:49]: Got it yes that makes sense!

File: op-challenger/fault/types.go [hunk 4]
```diff
@@ -19,19 +19,32 @@ type TraceProvider interface {
 	Get(i uint64) (common.Hash, error)
 }
 
-type Claim struct {
+// ClaimData is the core of a claim. It must be unique inside a specific game.
+
```
Comment (refcell) [inline op-challenger/fault/types.go:23]:
```suggestion
```

File: op-challenger/fault/types.go [hunk 5]
```diff
@@ -19,19 +19,32 @@ type TraceProvider interface {
...
+	Parent ClaimData
+}
+
+func (c *Claim) IsRoot() bool {
+	return c.Position.IsRootPosition()
+}
+
+func (c *Claim) DefendsParent() bool {
+	return (c.IndexAtDepth() >> 1) != c.Parent.IndexAtDepth()
 }
```
Comment (refcell) [inline op-challenger/fault/types.go:44]: NBD but having a docstring here would help make clear what defends means for a given claim

Comment (trianglesphere) [reply op-challenger/fault/types.go:44]: added

**Review by refcell — APPROVED** [review]

**Pull request merged**